### ISSUE TITLE
Can't use both TPresenter and TDoctrine traits

### ISSUE THREAD

**Issue by foxycode** [mrtnzlml/testbench#15]
If I use both, I get this error:

``` php
Fatal error: Trait method createContainer has not been applied, because there are collisions with other trait methods on Test\AdminModule\Presenters\ActuarialPresenterTest in /home/liquid/projects/ehub/portal/tests/AdminModule/presenters/ActuarialPresenterTest.phpt on line 12
```

Test looks like this:

``` php
class ActuarialPresenterTest extends \Tester\TestCase
{
    use \Testbench\TPresenter;
    use \Testbench\TDoctrine;

    public function setUp()
    {
        /** @var \Kdyby\Doctrine\EntityDao $userRepository */
        $userRepository = $this->getEntityManager()->getRepository(AdvertiserUser::class);
        $user = $userRepository->findOneBy(['advertiser.advertiserId' => 1]);
        $this->logIn($user);
    }

    public function testActionDefault()
    {
        $this->checkAction('Admin:Actuarial:default');
    }

}
```

**Comment by mrtnzlml** [comment]
This is known PHP limitation and I have no idea how can I solve this problem. See [this test case](https://github.com/mrtnzlml/testbench/blob/master/tests/Traits/MultipleTraits.phpt) (working workaround) and [this PHP bug report](https://bugs.php.net/bug.php?id=63911). If you have any idea how to solve this, please share your idea. Thanks...

**Comment by foxycode** [comment]
This is about common container functions i think. What about making use of `TCompiledContainer` mandatory? Then other traits will implement just it's own functionality.

**Comment by mrtnzlml** [comment]
This can lead to big troubles, because for example `TPresenter` depends on `TCompiledContainer` and I cannot simply remove this dependency form `TPresenter` and trust user that he is going to use `TCompiledContainer` trait (otherwise it will not work - weird). It's little bit more complicated.

You can also split your tests. Test presenter in one test case and databases in another. Or use workaround I sent before. Is this satisfactory?

I am thinking of inheritance instead of traits. Traits are great here, unfortunately it's not easy to use them with this bug (or behavior if you want to).

**Comment by foxycode** [comment]
I can't split tests. I use Doctrine for loading user identity from database to be able test admin presenters and your workaround is lot more complicated than faking that identity. Also because trait methods are private, I can't use workaround in base test class.

What about something like this:

``` php
trait TPresenter
{
    private function check($destination, $params = [], $post = [])
    {
        if (!is_callable([$this, 'getContainer'])) {
            trow new \Exception('Please use TCompiledContainer trair before TPresenter trair');
        }
        // ...
    }
}
```

And yes, inheritance would be definitely better suited for this case. I had it implemented ths way earlier:

``` php
<?php

namespace Test\AdminModule\Presenters;

use Nette;
use Tester\Assert;


$container = require __DIR__ . '/../../bootstrap.php';


class PayPresenterTest extends \Test\PresenterTestBase
{
    /**
     * @var Nette\Application\UI\Presenter
     */
    private $presenter;

    public function setUp()
    {
        $this->loginAdmin();
        $this->presenter = $this->createPresenter('Admin:Pay');
    }

    public function testActionDefault()
    {
        $request = new Nette\Application\Request('Admin:Pay', 'GET', [
            'action' => 'default',
        ]);

        $response = $this->presenter->run($request);

        Assert::true($response instanceof Nette\Application\Responses\TextResponse);
        Assert::true($response->getSource() instanceof Nette\Application\UI\ITemplate);
    }

}


(new PayPresenterTest($container))->run();
```

**Comment by mrtnzlml** [comment]
This is massive BC break even though there is only few users of this library. Please give me time to think about another implementations and possible solutions.

**Comment by foxycode** [comment]
@mrtnzlml Hope it won't take too much. I really need to refactor tests in one big app and I like to use testbench.

**Comment by mrtnzlml** [comment]
@foxycode Can you please try `dev-multiple-traits`? There should be possible to use multiple traits at once without big BC break (see [this test case](https://github.com/mrtnzlml/testbench/blob/86b0ef859a4b2dd43f2bf2610e64f10cbd70b3c4/tests/Traits/MultipleTraits.phpt)).

**EDIT**: if you need to use feature from `TCompiledContainer`, you have to name this trait explicitly in your test case (not needed before).

**Comment by foxycode** [comment]
@mrtnzlml I thinked about this a lot and I really liked more to use inheritance instead of traits so I could implement my custom logic on top of it. This way all traits need to be updated to provide protected functions.

Another thing is, I don't like the way container is created in this solution.

I can create short implementation and send PR if you want. Otherwise I'd reather create my own tester, even if it will be much inspired by yours :)

**Comment by mrtnzlml** [comment]
Can you please send me an example of what you are trying to do, but it's not possible? Creating DIC this way is perfectly fine...

**Comment by foxycode** [comment]
@mrtnzlml First of all, it's not possible make abstract test case, because I can't call trait funcitons from it's child.

**Comment by mrtnzlml** [comment]
Oh, this is stupid restriction I made before v2.0.0 release. I can make functions in traits protected again. It's should not be BC break. Would it be ok?

I do want to use traits, because it's perfect use case for traits and making this small change is really simple if it's helful. So why not... :)

Please try `dev-protected` ([example](https://github.com/mrtnzlml/testbench/blob/9cf9ac3a98532ecdd105b6fc1dcaa4411fd9ea8b/tests/Traits/TPresenterTest.phpt#L12))...

**Comment by foxycode** [comment]
@mrtnzlml It definitely won't hurt anyone. About other things I need to think more. Right now I need to concentrate on `Kdyby\Doctrine` rewrite. I'll definitely let you know :smile:

**Comment by mrtnzlml** [comment]
All fixed in [v2.0.4](https://github.com/mrtnzlml/testbench/releases/tag/v2.0.4)...
